File: get-comics.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"os/signal"
	"regexp"
	"strings"
	"syscall"
	"time"
	"bytes"
)

const default_config = "/usr/share/get-comics/comics.json"

var comics_dir = flag.String("d", "", "Output directory for comics")
var thread_limit = flag.Int("t", 4, "Maximum number of concurrent gets")
var gocomics_regexp regexp.Regexp
var gocomics_url string

var now = time.Now()
var wday = weekday2int()

// Counts
var total = 0
var got = 0
var skipped = 0

type Comic struct {
	// From the config file
	url       string
	host      string
	regexp    regexp.Regexp
	regmatch  int
	outname   string
	base_href string
	referer   string

	running   bool
}

var comics []Comic

func weekday2int() int {
	switch now.Weekday() {
	case time.Sunday:
		return 0
	case time.Monday:
		return 1
	case time.Tuesday:
		return 2
	case time.Wednesday:
		return 3
	case time.Thursday:
		return 4
	case time.Friday:
		return 5
	case time.Saturday:
		return 6
	default:
		return 0
	}
}

// This is not a complete implementation
func strftime(str string) string {
	if strings.Index(str, "%") != -1 {
		str = strings.Replace(str, "%Y", now.Format("2006"), -1)
		str = strings.Replace(str, "%m", now.Format("01"), -1)
		str = strings.Replace(str, "%d", now.Format("02"), -1)

		if strings.Index(str, "%") != -1 {
			panic("Bad time format: " + str)
		}
	}

	return str
}

func parse_comic(m map[string]interface{}, id int) {
	comic := make([]Comic, 1, 1)[0]

	for k, v := range m {
		switch val := v.(type) {
		case string:
			switch k {
			case "url":
				comic.url = strftime(val)

				u, err := url.Parse(comic.url)
				if err != nil {
					fmt.Println("Invalid url: ", comic.url)
					os.Exit(1)
				}
				comic.host = u.Scheme + "://" + u.Host
			case "days":
				if val[wday] == 'X' {
					skipped += 1
					return
				}
			case "regexp":
				val = strftime(val)
				comic.regexp = *regexp.MustCompile(val)
			case "output":
				comic.outname = val
			case "href":
				comic.base_href = val
				fmt.Println("Warning: We don't support href yet", comic.url)
				skipped += 1
				return
			case "referer":
				if val == "url" {
					comic.referer = comic.url
				} else {
					comic.referer = val
				}
			case "gocomic":
				comic.url = fmt.Sprintf(gocomics_url, val)
				comic.url = strftime(comic.url)
				comic.host = "http://www.gocomics.com"
				comic.regexp = gocomics_regexp
				comic.outname = val
			default:
				fmt.Println("Unexpected element ", k)
			}
		case float64:
			switch k {
			case "regmatch":
				comic.regmatch = int(val)
			case "redirect": /* redirects work silently */
			default:
				fmt.Println("Unexpected element ", k)
			}
		default:
			fmt.Println("Unkown element ", k)
		}
	}

	if comic.url == "" {
		fmt.Println("ERROR: comic with no url!")
		os.Exit(1)
	}

	comics = append(comics, comic)
	total += 1
}

func read_config(configfile string) {
	config, err := ioutil.ReadFile(configfile)
	if err != nil {
		fmt.Println(configfile, ": ", err)
		os.Exit(1)
	}

	// go json does not support comments... strip them
	state := 0
	nested := 0
	var b bytes.Buffer
	for i, c := range config {
		switch c {
		case '/':
			if state == 2 {
				state = 0
				continue
			}
			if config[i + 1] == '*' {
				state = 1
				nested += 1
				continue
			}
		case '*':
			if state == 1 {
				state = 0
				continue
			} else if nested > 0 && config[i + 1] == '/' {
				state = 2
				nested -= 1
				continue
			}
		}

		if nested == 0 {
			b.WriteByte(c)
		}
	}

	config = b.Bytes()

	var f interface{}
	err = json.Unmarshal(config, &f)
	if err != nil {
		fmt.Println("Unmarshal ", err)
		os.Exit(1)
	}

	m := f.(map[string]interface{})

	for k, v := range m {
		switch vv := v.(type) {
		case string:
			switch k {
			case "directory":
				/* Do not override the command line option */
				if *comics_dir == "" {
					*comics_dir = vv
				}
			case "gocomics-regexp":
				gocomics_regexp = *regexp.MustCompile(vv)
			case "gocomics-url":
				gocomics_url = vv
			default:
				fmt.Println("Unexpected element ", k)
			}
		case float64:
			switch k {
			case "threads":
				*thread_limit = int(vv)
			case "timeout":
				fmt.Println("Warning: timeout not supported.")
			case "randomize":
				fmt.Println("Warning: randomize not supported.")
			default:
				fmt.Println("Unexpected element ", k)
			}
		case []interface{}:
			if k == "comics" {
				for id, u := range vv {
					comic := u.(map[string]interface{})
					if len(comic) > 0 {
						parse_comic(comic, id)
					}
				}
			} else {
				fmt.Println("Unexpected element ", k)
			}
		default:
			fmt.Println("Unkown element ", k)
		}
	}

	if *comics_dir == "" {
		*comics_dir = os.Getenv("HOME") + "/comics"
	}
}

func dump_outstanding() {
	ch := make(chan os.Signal)
	signal.Notify(ch, syscall.SIGHUP)
	for {
		<-ch
		for i := range comics {
			if comics[i].running {
				fmt.Println("Running ", comics[i].url)
			}
		}
	}
}

/* This is a very lazy checking heuristic since we expect the files to
 * be one of the four formats and well formed. Yes, Close To Home
 * actually used TIFF. TIFF is only tested on little endian machine. */
func lazy_imgtype(hdr []byte) string {
	switch hdr[0] {
	case 'G':
		return ".gif"
	case 0xff:
		return ".jpg"
	case 0x89:
		return ".png"
	case 'M':
		return ".tif"
	case 'I':
		return ".tif"
	default:
		fmt.Println("Warning: Unknown file type: ", hdr)
		return ".xxx"
	}
}

func set_outname(comic Comic, hdr []byte) string {
	if comic.outname == "" { // Get file name from url
		ulen := len(comic.url)
		index := strings.LastIndex(comic.url, "/") + 1
		if index == 0 {
			comic.outname = comic.url
		} else if index >= ulen {
			comic.outname = "index.html"
		} else {
			comic.outname = comic.url[index:ulen]
		}
	}

	i := strings.LastIndex(comic.outname, ".")
	if i == -1 {
		comic.outname += lazy_imgtype(hdr)
	}

	return comic.outname
}

func gethttp(comic Comic, writeit bool) []byte {
	var resp *http.Response
	var err error

	if comic.referer == "" {
		resp, err = http.Get(comic.url)
	} else {
		client := &http.Client {}

		var req *http.Request
		req, err = http.NewRequest("GET", comic.url, nil)
		if err == nil {
			req.Header.Add("Referer", comic.referer)
			resp, err = client.Do(req)
		}
	}
	if err != nil {
		fmt.Println(comic.url, ": ", err)
		return nil
	}

	if resp.Status != "200 OK" {
		fmt.Println(comic.url, ": ", resp.Status)
		return nil
	}

	body, err := ioutil.ReadAll(resp.Body)
	resp.Body.Close()
	if err != nil {
		fmt.Println(comic.url, ": ", err)
		return nil
	}

	if !writeit {
		return body
	}

	outname := set_outname(comic, body[0:4])
	err = ioutil.WriteFile(outname, body, 0644)
	if err != nil {
		fmt.Println(outname, ": ", err)
	} else {
		got += 1
	}

	return nil
}

func find_match(comic Comic, body []byte) string {
	if comic.regmatch == 0 {
		matchb := comic.regexp.Find(body)
		if matchb == nil {
			fmt.Println(comic.url, ": No match.")
			return ""
		}

		return string(matchb)
	} else {
		matchb := comic.regexp.FindSubmatch(body)
		if matchb == nil {
			fmt.Println(comic.url, ": No match.")
			return ""
		}
		if len(matchb)-1 < comic.regmatch {
			fmt.Println(comic.url, ": No match for ", comic.regmatch)
			return ""
		}
		return string(matchb[comic.regmatch])
	}
}

func set_done(cur int, cs chan int) {
	comics[cur].running = false
	cs <- cur
}

func get_comic(cur int, cs chan int) {
	comics[cur].running = true
		defer set_done(cur, cs)

	comic := comics[cur]

	if comic.regexp.String() != "" {
		body := gethttp(comic, false)

		match := find_match(comic, body)
		if match == "" {
			return
		}

		if strings.HasPrefix(match, "http") {
			comic.url = match
		} else if match[0] == '/' {
			if match[1] == '/' {
				/* partially rooted - let's assume http */
				comic.url = "http:" + match
			} else {
				comic.url = comic.host + match
			}
		} else {
			comic.url = comic.host + "/" + match
		}
	}

	gethttp(comic, true)
}

func main() {
	flag.Parse()
	if len(flag.Args()) > 0 {
		for i := range flag.Args() {
			read_config(flag.Arg(i))
		}
	} else {
		read_config(default_config)
	}

	err := os.Chdir(*comics_dir)
	if err != nil {
		fmt.Println("Unable to change to " + *comics_dir)
		os.Exit(1)
	}

	cs := make(chan int)

	go dump_outstanding()

	// Start the first ones
	var cur int
	for cur = 0; cur < *thread_limit && cur < total; cur++ {
		go get_comic(cur, cs)
	}

	for alldone := 0; alldone < total; alldone++ {
		<-cs // block waiting for a comic to finish

		if cur < total {
			go get_comic(cur, cs)
			cur += 1
		}
	}

	fmt.Printf("Got %d of %d (%d skipped)\n", got, total, skipped)
}

File: get-comics.py
#!/usr/bin/python

import sys, os, re, json, threading, argparse, time, random, signal
from datetime import date

# You probably don't have requests. First make sure you have pip, if not:
#    sudo python -m ensurepip
# Then grab requests and install it:
#    git clone git://github.com/requests/requests.git
#    cd requests
#    sudo pip install .
import requests

comment_re = re.compile(r"/\*(.*?)\*/")
multi_comment_re = re.compile(r"/\*(.*?)\*/", re.DOTALL)

comics_dir = None
index_dir = None
links_only = None
links_lock = None

comics = list()
nthreads = 10
threads = list()

class Comic:
    gocomics_regexp = None
    weekday = 0
    today = 0
    __id = 0

    # stats
    total = 0
    got = 0
    skipped = 0

    def __init__(self):
        self.url = None
        self.host = None
        self.regexp = None
        self.regmatch = 0
        self.skip = False
        self.state = 0
        # Make sure every comic has an outname
        self.outname = "comic" + str(Comic.__id)
        Comic.__id += 1
        Comic.total += 1

    def add_url(self, url):
        self.url = time.strftime(url, Comic.today)
        m = re.match(r"https?://[^/]+", url)
        if m:
            self.host = m.group(0)
        else:
            self.host = ""
            print "ERROR: Unable to isolate host: " + url

    def add_regexp(self, regexp):
        self.regexp = time.strftime(regexp, Comic.today)

    def add_days(self, days):
        if days[Comic.weekday] == 'X':
            self.skip = True

    def add_gocomic(self, comic):
        if self.gocomics_regexp == None:
            print "PROBLEMS: No gocomics_regexp"
            sys.exit(1)
        self.url = "http://www.gocomics.com/" + comic + "/"
        self.regexp = self.gocomics_regexp
        self.outname = comic

def strip_comments (data):
    # First remove all single line comments to solve nested comment
    # problems
    while True:
        m = re.search(comment_re, data)
        if m == None:
            break
        data = data[:m.start()] + data[m.end():]

    # Now remove the multiline comments
    while True:
        m = re.search(multi_comment_re, data)
        if m == None:
            break
        data = data[:m.start()] + data[m.end():]

    return data

def parse_comic (comic):
    new_comic = Comic()
    comics.append(new_comic)
    for each in comic:
        if each == 'url':
            new_comic.add_url(comic['url'])
        elif each == 'days':
            new_comic.add_days(comic['days'])
        elif each == 'regexp':
            new_comic.add_regexp(comic['regexp'])
        elif each == 'regmatch':
            new_comic.regmatch = comic['regmatch']
        elif each == 'output':
            new_comic.outname = comic['output']
        elif each == 'referer':
            pass
        elif each == 'gocomic':
            new_comic.add_gocomic(comic['gocomic'])
        else:
            print("Unexpected key " + each)

def read_config (fname):
    global comics_dir

    try:
        fp = open(fname)
        data = fp.read()
        fp.close()
    except Exception,e:
        print(e.args[1] + ": " + fname)
        sys.exit(1)

    # Python json cannot handle comments
    data = strip_comments(data)

    data = json.loads(data)

    try: Comic.gocomics_regexp = data['gocomics-regexp']
    except: pass

    # Get the weekday and make Sunday 0
    Comic.today = time.localtime();
    Comic.weekday = date.today().weekday() + 1
    if Comic.weekday == 7: Comic.weekday = 0

    for each in data:
        if each == 'comics':
            for comic in data['comics']:
                if len(comic) > 0:
                    parse_comic(comic)
        elif each == 'threads':
            nthreads = data['threads']
        elif each == 'directory':
            comics_dir = data['directory']
        elif each != 'gocomics-regexp':
            print "WARNING: " + str(each)

imgtype = { 'GIF8':             '.gif',
            '\x89PNG':          '.png',
            '\xff\xd8\xff\xe0': '.jpg', # jfif
            '\xff\xd8\xff\xe1': '.jpg', # exif
            '\xff\xd8\xff\xee': '.jpg', # Adobe
            'II\x42\0':         '.tif', # little endian
            'MM\0\x42':         '.tif' # little endian
            }

file_exts = { '.gif', '.png', '.jpg', '.tif', '.xxx', '.html' }

# This is a very lazy checking heuristic since we expect the files to
# be one of the four formats and well formed. Yes, Close To Home
# actually used TIFF. TIFF is only tested on little endian machines.
def lazy_imgtype (content):
    # print ':'.join(x.encode('hex') for x in content[0:4])
    # print content[0:4]
    try: return imgtype[content[0:4]]
    except: return '.xxx'

def clean_directory (dir):
    for file in os.listdir(dir):
        if os.path.splitext(file)[1] in file_exts:
            os.remove(dir + '/' + file)
        else:
            print 'Warning: Not cleaning ' + file

def handler(signum, frame):
    print "Current state:"
    for comic in comics:
        if not (comic.state == 2 or comic.skip):
            if comic.state == 1:
                print "  Running " + comic.url
            else:
                print "  Queued  " + comic.url

def outfile (comic, text, addext = True, dir = None):
    global comics_dir

    if dir == None: dir = comics_dir

    if addext:
        ext = lazy_imgtype(text)
    else:
        ext = ""
    try:
        fp = open(dir + '/' + comic.outname + ext, 'w')
        fp.write(text)
        fp.close()
        return True
    except Exception,e:
        print str(e.args[1]) + ": " + comic.outname
        return False

def stage2 (comic, url):
    if url.startswith("http"):
        pass
    elif url.startswith("//"):
        url = "http:" + url
    else:
        url = comic.host + "/" + url

    if links_only:
        links_lock.acquire()
        links_only.write(url + '\n')
        links_lock.release()
        Comic.got += 1
        return

    r = requests.get(url)
    if r.status_code != 200:
        print "Error:" + url + " " + str(r.status_code)
        return
    if outfile(comic, r.content):
        Comic.got += 1

def comic_thread (comic):
    global sema

    comic.state = 1 # Running
    r = requests.get(comic.url)
    if r.status_code == 200:
        if comic.regexp:
            m = re.search(comic.regexp, r.text)
            if m:
                stage2(comic, m.group(comic.regmatch))
            else:
                print "ERROR: " + comic.url + " did not match regexp"
                # Save the file for debugging
                comic.outname = comic.outname + ".html"
                outfile(comic, r.text.encode('ascii', 'ignore'), False, index_dir)
        elif links_only:
            links_lock.acquire()
            links_only.write(comic.url + '\n')
            links_lock.release()
            Comic.got += 1
        else:
            if outfile(comic, r.content):
                Comic.got += 1
    else:
        print "Error:" + comic.url + " " + str(r.status_code)

    comic.state = 2 # Done
    sema.release()

def comic_thread2 (comic): # Dummy version for testing
    global sema

    time.sleep(random.randint(1, 4))
    sema.release()

### Main

parser = argparse.ArgumentParser()
parser.add_argument('-c', action='store_true', help='clean comics directory')
parser.add_argument('-d', metavar='comics dir', help='comics directory')
parser.add_argument('-i', metavar='index dir', help='index directory')
parser.add_argument('-l', metavar='links file', help='links file')
parser.add_argument('-v', action='count', help='verbose')
parser.add_argument('config', nargs='?', help='config file', default="/usr/share/get-comics/comics.json")
args = parser.parse_args()

read_config(args.config)

# Setup comics_dir and make sure it exists
if args.d:
    comics_dir = args.d
elif comics_dir == None:
    comics_dir = os.getenv('HOME') + '/comics'
if not os.path.isdir(comics_dir):
    print "ERROR: " + comics_dir + " does not exist"
    sys.exit(1)

# Setup index_dir and make sure it exists
if args.i:
    index_dir = args.i
else:
    index_dir = comics_dir
if not os.path.isdir(index_dir):
    print "ERROR: " + index_dir + " does not exist"
    sys.exit(1)

if args.c:
    clean_directory(comics_dir)
    if args.i:
        clean_directory(index_dir)

if args.l:
    try:
        links_only = open(args.l, 'w');
    except Exception,e:
        print str(e.args[1]) + ": " + args.l
        sys.exit(1)
    links_lock = threading.Lock()

# We use a semaphore to limit the number of outstanding threads
sema = threading.Semaphore(nthreads)

signal.signal(signal.SIGUSR1, handler)

for comic in comics:
    if comic.skip:
        Comic.skipped += 1
    else:
        sema.acquire()
        thread = threading.Thread(target=comic_thread, args=(comic, ))
        threads.append(thread)
        if args.v: print "Starting " + thread.name + " " + comic.url
        thread.start()

for thread in threads:
    if comic.skip == False:
        thread.join()
        if args.v: print "Done " + thread.name

if links_only:
    links_only.close()

print "Got " + str(Comic.got) + " skipped " + str(Comic.skipped) + " of " + str(Comic.total)
